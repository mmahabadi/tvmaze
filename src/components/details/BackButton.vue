<script setup lang="ts">
import { useRouter } from 'vue-router'
import BackIcon from '@/@ui/components/icons/BackIcon.vue'
const router = useRouter()

const goBack = () => {
  router.go(-1)
}
</script>
<template>
  <button @click="goBack" class="btn-back">
    <div class="text">
      Back
      <BackIcon class="icon" />
    </div>
  </button>
</template>
<style>
.btn-back {
  @apply lg:w-1/4 hidden  w-full xl:flex justify-center items-center hover:bg-transparent border-2 transition duration-300 md:h-64 h-20 rounded font-medium;

  background-color: var(--primary-color);
  border-color: var(--primary-color);

  .text {
    @apply flex justify-center items-center gap-6 text-xl uppercase tracking-widest lg:rotate-90;
  }
  .icon {
    @apply h-9 w-9 md:-rotate-90;
  }
}
</style>
