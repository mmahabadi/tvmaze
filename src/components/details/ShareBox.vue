<script setup lang="ts">
import { defineProps } from 'vue'
import ShareIcon from '@/@ui/components/icons/ShareIcon.vue'
import IconPlay from '@/@ui/components/icons/IconPlay.vue'

defineProps<{
  language: string
}>()
</script>
<template>
  <div class="box">
    <div class="col1">
      <button class="share-btn">
        <ShareIcon />
      </button>
    </div>
    <div class="col2">
      <p>
        Language : <span class="ml-2 truncate">{{ language }}</span>
      </p>
    </div>
    <div class="col3">
      <button class="play-btn">
        <IconPlay />
        Watch
      </button>
    </div>
  </div>
</template>
<style scoped>
.box {
  @apply grid sm:grid-cols-5 grid-cols-3 gap-4 p-6 items-center bg-gray-900 border border-gray-800 rounded-lg;

  .col1 {
    @apply col-span-1 border-r hidden sm:flex;
  }
  .col2 {
    @apply col-span-2 flex-col hidden sm:flex font-medium text-sm;
  }
  .col3 {
    @apply sm:col-span-2 col-span-3 flex justify-end font-medium text-sm;
  }

  .share-btn {
    @apply w-10 h-10 flex justify-center items-center rounded-lg bg-white bg-opacity-20;
  }
  .play-btn {
    @apply py-4 transition border-2 rounded-full flex justify-center items-center gap-4 w-full sm:py-3;
    border-color: var(--primary-color);
  }
  .share-btn:hover,
  .play-btn:hover {
    background-color: var(--primary-color);
  }
}
</style>
