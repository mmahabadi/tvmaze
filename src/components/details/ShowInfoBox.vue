<script setup lang="ts">
import { defineProps } from 'vue'
import CalendarIcon from '@/@ui/components/icons/CalendarIcon.vue'
import ClockIcon from '@/@ui/components/icons/ClockIcon.vue'
import { Movie } from '@/types'

type propType = Partial<Pick<Movie, 'status' | 'year' | 'days' | 'time'>> & { genres?: string }

defineProps<propType>()
</script>
<template>
  <div class="box">
    <div class="status">{{ status }}</div>
    <div class="item">
      <span class="text">{{ genres }}</span>
    </div>
    <div class="item">
      <CalendarIcon class="text-primary" />
      <span class="text">
        {{ year }}
      </span>
    </div>
    <div class="item">
      <ClockIcon class="text-primary" />
      <span class="text">{{ days }} {{ time }}</span>
    </div>
  </div>
</template>
<style>
.box {
  @apply flex items-center gap-4 font-medium;

  .status {
    @apply text-xs px-2 py-1;
    background-color: var(--primary-color);
  }

  .item {
    @apply flex items-center gap-2;

    .text-primary {
      color: var(--primary-color);
    }
    .text {
      @apply text-xs font-medium;
    }
  }
}
</style>
