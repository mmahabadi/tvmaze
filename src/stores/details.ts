import { defineStore } from 'pinia'
import { ref } from 'vue'
import * as ApiService from '@/services/ApiService'
import type { Movie } from '@/types'

export const useDetailsStore = defineStore('details', () => {
  const details = ref<Movie | null>(null)
  const loading = ref(false)
  const error = ref<null | Error>(null)

  async function getShowDetails(id: number): Promise<void> {
    loading.value = true
    try {
      details.value = await ApiService.getDetails(id)
    } catch (e) {
      error.value = e as Error
    } finally {
      loading.value = false
    }
  }

  return {
    details,
    loading,
    error,
    getShowDetails
  }
})
