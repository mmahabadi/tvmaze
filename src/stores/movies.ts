import { defineStore } from 'pinia'
import { computed, ref, type ComputedRef } from 'vue'
import * as ApiService from '@/services/ApiService'
import { type Movie, type ShowByGenre } from '@/types'

export const useMoviesStore = defineStore('movies', () => {
  const shows = ref<Map<string, ShowByGenre>>(new Map())
  const movies = ref<Movie[]>([])
  const details = ref<Movie | null>(null)
  const loading = ref(false)
  const error = ref<null | Error>(null)
  const currentIndex = ref(1)
  const ITEMS_PER_PAGE = 3
  const ITEMS_LIMIT_FOR_RECENT_MOVIES = 15
  const DATE_LIMIT_FOR_RECENT_MOVIES = 2020

  const recentMovies = computed(() =>
    movies.value
      .filter((movie) => new Date(movie.ended).getFullYear() > DATE_LIMIT_FOR_RECENT_MOVIES)
      .slice(0, ITEMS_LIMIT_FOR_RECENT_MOVIES)
  )

  /**
   * @note because the API does not support fetching genres,
   * I need to extract the genres from the fetched movies
   */
  const genres = computed(() => [...new Set([...movies.value.flatMap((movie) => movie.genres)])])

  const getByGenre = computed(() => (genre: string): ShowByGenre => {
    loadGenre(genre)
    return shows.value.get(genre) as ShowByGenre
  })

  const paginatedGenres: ComputedRef<string[]> = computed(() => {
    const startIndex = currentIndex.value
    const endIndex = startIndex * ITEMS_PER_PAGE
    // Concatenate the new slice with the previous values of pagedGenres
    return [...genres.value.slice(0, endIndex)]
  })

  async function fetchMovies(): Promise<void> {
    loading.value = true
    try {
      movies.value = await ApiService.getData()
    } catch (e) {
      error.value = e as Error
    } finally {
      loading.value = false
    }
  }

  async function fetchShowByGenre(genre: string): Promise<void> {
    const genreShows = shows.value.get(genre)
    if (!genreShows) return

    try {
      /**
       * @note if the APi supports fetching movies by genre,
       * comment the following line
       * and uncomment the line below
       */
      const items = movies.value
        .filter((movie) => movie.genres.includes(genre))
        .sort((a, b) => b.rating.average - a.rating.average)
      /**
       * @note the API does not support fetching movies by genre
       * uncomment the following line to fetch the movies by genre from the API
       */
      // const items = await ApiService.getShowsByGenre(genre)
      if (items) {
        genreShows.movies = items
      }
    } catch (e) {
      genreShows.error = e as Error
    } finally {
      genreShows.loading = false
    }
    shows.value.set(genre, genreShows as ShowByGenre)
  }

  function loadGenre(genre: string): void {
    if (!shows.value.has(genre)) {
      const genreShows: ShowByGenre = {
        genre,
        movies: [],
        loading: true,
        error: null
      }
      shows.value.set(genre, genreShows)
      fetchShowByGenre(genre)
    }
  }

  function nextPage(): void {
    if (currentIndex.value * ITEMS_PER_PAGE <= genres.value.length) currentIndex.value++
  }

  async function getShowDetails(id: number): Promise<void> {
    loading.value = true
    try {
      details.value = await ApiService.getDetails(id)
    } catch (e) {
      error.value = e as Error
    } finally {
      loading.value = false
    }
  }

  return {
    movies,
    genres: paginatedGenres,
    recentMovies,
    details,
    loading,
    error,
    fetchMovies,
    getByGenre,
    nextPage,
    getShowDetails
  }
})
