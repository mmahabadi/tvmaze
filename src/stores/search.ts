import type { Movie } from '@/types'
import { defineStore } from 'pinia'
import { ref } from 'vue'
import * as ApiService from '@/services/ApiService'

export const useSearchStore = defineStore('search', () => {
  const searchResults = ref<Movie[]>([])
  let searchQuery = ''
  const loading = ref(false)
  const error = ref<null | Error>(null)

  async function searchShows(query: string): Promise<void> {
    //avoid making the same request multiple times
    if (searchQuery === query) return
    searchQuery = query
    loading.value = true
    try {
      searchResults.value = await ApiService.search(query)
    } catch (e) {
      error.value = e as Error
    } finally {
      loading.value = false
    }
  }

  return {
    searchResults,
    searchQuery,
    loading,
    error,
    searchShows
  }
})
