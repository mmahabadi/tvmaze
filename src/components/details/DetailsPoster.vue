<script setup lang="ts">
import DOMPurify from 'dompurify'
import { computed, defineProps } from 'vue'
import type { Movie } from '@/types'
import Image from '@/@ui/components/Image.vue'
import BackButton from './BackButton.vue'
import ShowInfoBox from './ShowInfoBox.vue'
import ShareBox from './ShareBox.vue'

const props = defineProps<{ item: Movie }>()
const { item } = props

const year = computed(() => {
  return item.ended?.split('-')[0]
})

const sanitizedSummary = computed(() => DOMPurify.sanitize(item.summary))
</script>
<template>
  <div data-testid="poster" class="poster-container">
    <Image :src="item.image?.original" :alt="item.name" class="large-image" />
    <div class="overlay">
      <div class="container">
        <div data-testid="thumbnail" class="image-content">
          <Image :src="item.image?.medium" :alt="item.name" class="thumbnail" />
        </div>
        <div class="main-content">
          <div class="inner-container">
            <h1 class="title">
              {{ item.name }}
            </h1>
            <ShowInfoBox
              :genres="item.genres.join(',')"
              :status="item.status"
              :days="item.schedule.days?.join(',')"
              :year="year"
              :time="item.schedule.time"
            />
            <p class="summary" v-html="sanitizedSummary"></p>
            <ShareBox :language="item.language" />
          </div>
          <div class="back-container">
            <BackButton />
          </div>
        </div>
      </div>
    </div>
  </div>
</template>
<style scoped>
.poster-container {
  @apply w-full h-full xl:max-h-[44rem] xl:overflow-hidden relative text-white;

  .large-image {
    @apply w-full hidden xl:inline-block h-full object-cover object-top;
  }
  .overlay {
    @apply xl:bg-black/90 flex-col xl:bg-opacity-90 xl:absolute top-0 left-0 right-0 bottom-0;

    .container {
      @apply px-3 mx-auto 2xl:px-32 flex-col items-center justify-center lg:grid grid-cols-3 py-10 lg:py-20 gap-8;
    }

    .image-content {
      @apply col-span-3 md:col-span-1 mx-auto w-[12rem] md:block lg:w-full border border-gray-800 rounded-lg overflow-hidden;

      .thumbnail {
        @apply xl:w-full xl:h-full xl:object-cover;
      }
    }
  }

  .main-content {
    @apply col-span-2 lg:grid grid-cols-5 gap-4 items-center;

    .inner-container {
      @apply col-span-5 xl:col-span-3 flex flex-col gap-10;

      .title {
        @apply xl:text-4xl capitalize font-sans text-2xl font-bold;
      }
    }

    .summary {
      @apply text-sm leading-7;
    }
    .back-container {
      @apply col-span-2 md:mt-0 mt-2 flex justify-center items-center;
    }
  }
}
</style>
