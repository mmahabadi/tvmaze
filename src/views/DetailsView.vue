<script setup lang="ts">
import { computed, onMounted, watch } from 'vue'
import { useRoute } from 'vue-router'
import { useMoviesStore } from '@/stores/movies'
import MainLayout from '@/layouts/MainLayout.vue'
import DetailsPoster from '@/components/details/DetailsPoster.vue'
import CasteSwiper from '@/components/casts/CastSwiper.vue'
import { type Person } from '@/types'

const route = useRoute()
const store = useMoviesStore()
const showId = computed(() => +route.params.id)

watch(
  () => route.params.id,
  (newValue) => {
    getShow(+newValue)
  }
)

onMounted(() => {
  getShow(showId.value)
})

const getShow = (id: number) => {
  store.getShowDetails(id)
}

const casts = computed(() => store.details?._embedded?.cast?.map((item) => item.person as Person))
</script>

<template>
  <MainLayout :loading="store.loading">
    <template v-slot:slider>
      <DetailsPoster v-if="store.details" :item="store.details" />
    </template>
    <CasteSwiper v-if="!!casts?.length" :slides="casts" />
  </MainLayout>
</template>
