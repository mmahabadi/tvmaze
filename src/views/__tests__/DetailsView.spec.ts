import { describe, it, expect, beforeEach, vi, type MockInstance } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import DetailsView from '../DetailsView.vue'
import { createPinia, setActivePinia } from 'pinia'
import { useRoute } from 'vue-router'
import { useDetailsStore } from '@/stores/details'

vi.mock('vue-router', () => ({
  useRoute: vi.fn()
}))

const mockRoute = {
  params: {
    id: 1
  }
}

describe('DetailsView', () => {
  let wrapper: VueWrapper<DetailsView>
  let fetchDetailsSpy: MockInstance<[id: number], Promise<void>>

  beforeEach(() => {
    setActivePinia(createPinia())
    ;(useRoute as any).mockReturnValue(mockRoute)

    const store = useDetailsStore()
    fetchDetailsSpy = vi.spyOn(store, 'getShowDetails')

    wrapper = mount(DetailsView, {
      global: {
        stubs: {
          MainLayout: true,
          DetailsPoster: true,
          CasteSwiper: true
        },
        provide: {
          store
        }
      }
    })
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('calls fetchMovie on mount', () => {
    //Assert
    expect(fetchDetailsSpy).toHaveBeenCalledWith(mockRoute.params.id)
  })
})
