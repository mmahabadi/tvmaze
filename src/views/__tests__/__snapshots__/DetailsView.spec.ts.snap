// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DetailsView > renders correctly 1`] = `"<main-layout-stub loading="true"></main-layout-stub>"`;
